### PR TITLE
feat(7481): Operator to ensure an expression is contextually typed by, and satisfies, some type

### LINKED ISSUE #7481: Operator to ensure an expression is contextually typed by, and satisfies, some type
Sometimes it's necessary (e.g. for guiding type inference, for ensuring sub-expression conforms to an interface, or for clarity) to change the static type of an expression. Currently TypeScript has the `as` (aka `<>`) operator for that, but it's dangerous, as it also allows down-casting. It would be nice if there was another operator for implicit conversions only (type compatibility). I think this operator should be recommended in most cases instead of `as`.

This operator can be implemented as a generic function, but as it shouldn't have any run-time effect, it would be better if it was incorporated into the language.

``` typescript
function asType<T>(value: T) {
  return value;
};
```

EDIT: Due to parameter bivariance, this function is _not_ equivalent to the proposed operator, because asType allows downcasts too.

### LINKED ISSUE #47920: "satisfies" operator to ensure an expression matches some type (feedback reset)
# Feature Update - February 2022

This is a feedback reset for #7481 to get a fresh start and clarify where we are with this feature. I really thought this was going to be simpler, but it's turned out to be a bit of a rat's nest!

Let's start with what kind of scenarios we think need to be addressed.

## Scenario Candidates

First, here's a review of scenarios I've collected from reading the linked issue and its many duplicates. **Please post if you have other scenarios that seem relevant**. I'll go into it later, but *not all these scenarios can be satisfied at once* for reasons that will hopefully become obvious.

### Safe Upcast

Frequently in places where control flow analysis hits some limitation (hi #9998), it's desirable to "undo" the specificity of an initializer. A good example would be
```ts
let a = false;
upd();
if (a === true) {
//    ^^^ error, true and false have no overlap
    // ...
}
function upd() {
    if (someCondition) a = true;
}
```
The canonical recommendation is to type-assert the initializer:
```ts
let a = false as boolean;
```
but there's limited type safety here since you could accidently downcast without realizing it:
```ts
type Animal = { kind: "cat", meows: true } | { kind: "dog", barks: true };
let p = { kind: "cat" } as Animal; // Missing meows!
upd();
if (p.kind === "dog") {

} else {
    p.meows; // Reported 'true', actually 'undefined'
}
function upd() {
    if (Math.random() > 0.5) p = { kind: "dog", barks: true };
}
```

The safest workaround is to have a dummy function, `function up<T>(arg: T): T`:
```ts
let a = up<boolean>(true);
```
which is unfortunate due to having unnecessary runtime impact.

Instead, we would presumably write
```ts
let p = { kind: "cat", meows: true } satisfies Animal;
```

### Property Name Constraining

We might want to make a lookup table where the property keys must come from some predefined subset, but not lose type information about what each property's value was:
```ts
type Keys = 'a' | 'b' | 'c' | 'd';

const p = {
    a: 0,
    b: "hello",
    x: 8 // Should error, 'x' isn't in 'Keys'
};

// Should be OK -- retain info that a is number and b is string
let a = p.a.toFixed();
let b = p.b.substr(1);
// Should error even though 'd' is in 'Keys'
let d = p.d;
```
There is no obvious workaround here today.

Instead, we would presumably write
```ts
const p = {
    a: 0,
    b: "hello",
    x: 8 // Should error, 'x' isn't in 'Keys'
} satisfies Partial<Record<Keys, unknown>>;
// using 'Partial' to indicate it's OK 'd' is missing
```

### Property Name Fulfillment

Same as Property Name Constraining, except we might want to ensure that we get *all* of the keys:

```ts
type Keys = 'a' | 'b' | 'c' | 'd';

const p = {
    a: 0,
    b: "hello",
    c: true
    // Should error because 'd' is missing
};
// Should be OK
const t: boolean = p.c;
```

The closest available workaround is:
```ts
const dummy: Record<Keys, unknown> = p;
```
but this assignment a) has runtime impact and b) will not detect excess properties.

Instead, we would presumably write
```ts
const p = {
    a: 0,
    b: "hello",
    c: true
    // will error because 'd' is missing
} satisfies Record<Keys, unknown>;
```

### Property Value Conformance

This is the flipside of Property Name Constraining - we might want to make sure that all *property values* in an object conform to some type, but still keep record of which keys are present:
```ts
type Facts = { [key: string]: boolean };
declare function checkTruths(x: Facts): void;
declare function checkM(x: { m: boolean }): void;
const x = {
    m: true
};

// Should be OK
checkTruths(x);
// Should be OK
fn(x);
// Should fail under --noIndexSignaturePropertyAccess
console.log(x.z);
// Should be OK under --noUncheckedIndexedAccess
const m: boolean = x.m;

// Should be 'm'
type M = keyof typeof x;

// Should be able to detect a failure here
const x2 = {
    m: true,
    s: "false"
};
```

Another example
```ts
export type Color = { r: number, g: number, b: number };

// All of these should be Colors, but I only use some of them here.
export const Palette = {
    white: { r: 255, g: 255, b: 255},
    black: { r: 0, g: 0, d: 0}, // <- oops! 'd' in place of 'b'
    blue: { r: 0, g: 0, b: 255 },
};
```

Here, we would presumably write
```ts
const Palette = {
    white: { r: 255, g: 255, b: 255},
    black: { r: 0, g: 0, d: 0}, // <- error is now detected
    blue: { r: 0, g: 0, b: 255 },
} satisfies Record<string, Color>;
```

### Ensure Interface Implementation

We might want to leverage type inference, but still check that something conforms to an interface and use that interface to provide contextual typing:
```ts
type Movable = {
    move(distance: number): void;
};

const car = {
    start() { },
    move(d) {
        // d should be number
    },
    stop() { }
};
```

Here, we would presumably write
```ts
const car = {
    start() { },
    move(d) {
        // d: number
    },
    stop() { }
} satisfies Moveable;
```

### Optional Member Conformance

We might want to initialize a value conforming to some weakly-typed interface:
```ts
type Point2d = { x: number, y: number };
// Undesirable behavior today with type annotation
const a: Partial<Point2d> = { x: 10 };
// Errors, but should be OK -- we know x is there
console.log(a.x.toFixed());
// OK, but should be an error -- we know y is missing
let p = a.y;
```

### Optional Member Addition

Conversely, we might want to safely initialize a variable according to some type but retain information about the members which aren't present:
```ts
type Point2d = { x: number, y: number };
const a: Partial<Point2d> = { x: 10 };
// Should be OK
a.x.toFixed();
// Should be OK, y is present, just not initialized
a.y = 3;
```

### Contextual Typing

TypeScript has a process called *contextual typing* in which expressions which would otherwise not have an inferrable type can get an inferred type from context:
```ts
//         a: implicit any
const f1 = a => { };

//                              a: string
const f2: (s: string) => void = a => { };
```

In all of the above scenarios, contextual typing would always be appropriate. For example, in *Property Value Conformance*
```ts
type Predicates = { [s: string]: (n: number) => boolean };

const p: Predicates = {
    isEven: n => n % 2 === 0,
    isOdd: n => n % 2 === 1
};
```
Contextually providing the `n` parameters a `number` type is clearly desirable. In most other places than parameters, the contextual typing of an expression is not directly observable except insofar as normally-disallowed assignments become allowable.

## Desired Behavior Rundown

There are three plausible contenders for what to infer for the type of an `e satisfies T` expression:
 * `typeof e`
 * `T`
 * `T & typeof e`

*SATA: Same As Type Annotation - `const v = e satisfies T` would do the same as `const v: T = e`, thus no additional value is provided

|Scenario|`T`|`typeof e`|`T & typeof e`|
|--------|---|----------|--------------|
|Safe Upcast|✔|❌ (undoes the upcasting)|❌ (undoes the upcasting)|
|Property Name Constraining|❌ (SATA)|✔|✔|
|Property Name Fulfillment|❌ (SATA)|✔|✔|
|Ensure Interface Implementation|❌ (SATA)|✔|✔|
|Optional Member Conformance|❌ (SATA)|✔|❌ (members appear when not desired)|
|Optional Member Addition|❌ (SATA)|❌ (members do not appear when desired)|✔|
|Contextual Typing|✔|✔|✔|

### Discussion

Given the value of the other scenarios, I think *safe upcast* needs to be discarded. One could imagine other solutions to this problem, e.g. marking a particular variable as "volatile" such that narrowings no longer apply to it, or simply by having better side-effect tracking.

### Excess Properties

A sidenote here on excess properties. Consider this case:
```ts
type Point = {
    x: number,
    y: number
};
const origin = {
    x: 0,
    y: 0,
    z: 0 // OK or error?
} satisifes Point;
```

Is `z` an excess property?

One argument says yes, because in other positions where that object literal was used where a `Point` was expected, it would be. Additionally, if we want to detect typos (as in the *property name constraining* scenario), then detecting excess properties is mandatory.

The other argument says no, because the point of excess property checks is to detect properties which are "lost" due to not having their presence captured by the type system, and the design of the `satisfies` operator is specifically for scenarios where the ultimate type of all properties *is* captured somewhere.

I think on balance, the "yes" argument is stronger. If we don't flag excess properties, then the *property name constraining* scenario can't be made to work at all. In places where excess properties are expected, `e satisfies (T & Record<string, unknown>)` can be written instead.

However, under this solution, producing the expression type `T & typeof e` becomes very undesirable:
```ts
type Point2d = { x: number, y: number };
const a = { x: 10, z: 0 } satisfies Partial<Point2d> & Record<string, unknown>;
// Arbitrary writes allowed (super bad)
a.blah = 10;
```

Side note: It's tempting to say that properties aren't excess if all of the satisfied type's properties are matched. I don't think this is satisfactory because it doesn't really clearly define what would happen with the asserted-to type is `Partial`, which is likely common:
```ts
type Keys = 'a' | 'b' | 'c';
// Property 'd' might be intentional excess *or* a typo of e.g. 'b'
const v = { a: 0, d: 0 } satisfies Partial<Record<Keys, number>>;
```

Producing `typeof e` then leads to another problem...

### The Empty Array Problem

Under `--strict` (specifically `strictNullChecks && noImplicitAny`), empty arrays in positions where they can't be Evolving Arrays get the type `never[]`. This leads to some somewhat annoying behavior today:
```ts
let m = { value: [] };
// Error, can't assign 'number' to 'never'
m.value.push(3);
```

The `satisfies` operator might be thought to fix this:
```ts
type BoxOfArray<T> = { value: T[] };
let m = { value: [] } satisfies BoxOfArray<number>
// OK, right? I just said it was OK?
m.value.push(3);
```

However, under current semantics (including `m: typeof e`), this still doesn't work, because the type of the array is still `never[]`.

It seems like this can be fixed with a targeted change to empty arrays, which I've prototyped at #47898. It's possible there are unintended downstream consequences of this (changes like this can often foul up generic inference in ways that aren't obvious), but it seems to be OK for now.

## TL;DR

It seems like the best place we could land is:
 * Contextually type empty arrays
 * Disallow excess properties (unless you write `T & Record<string, unknown>`)
 * Use `typeof e` as the expression type instead of `T` or `T & typeof e`

Does this seem right? What did I miss?

### PR DESCRIPTION
**Pull request by a-tarasyuk**
Fixes #7481
Fixes #47920

### CODE REVIEW

**Comment by RyanCavanaugh** [comment]
@typescript-bot pack this

**Comment by RyanCavanaugh** [comment]
Let's try again 

@typescript-bot pack this

**Comment by andrewbranch** [comment]
@typescript-bot pack this

File: src/compiler/checker.ts
Comment (RyanCavanaugh) [inline src/compiler/checker.ts:31331]: We need a new case in `getContextualType`. This code shouldn't have an implicit any:
```ts
type MyFunc = (x: string) => void;
const f = (m => m.substr(0)) satisfies MyFunc;
```

**Review by RyanCavanaugh — DISMISSED** [review]

File: src/compiler/checker.ts
Comment (RyanCavanaugh) [inline src/compiler/checker.ts:31329]: Per discussion at the issue, we should return `exprType` here, not `targetType`

File: src/compiler/diagnosticMessages.json
Comment (DanielRosenwasser) [inline src/compiler/diagnosticMessages.json:1063]: Type '{0}' does not satisfy the expected type '{1}'.

File: src/compiler/types.ts
Comment (DanielRosenwasser) [inline src/compiler/types.ts:7035]: I wonder if this needs to be different in any way than an `TypeAssertions`. I'm kind of surprised that we expose this enum publicly.

**Comment by a-tarasyuk** [comment]
@DanielRosenwasser @RyanCavanaugh Thanks for the feedback. I've added the requested changes.

**Comment by RyanCavanaugh** [comment]
@typescript-bot pack this

**Comment by rbuckton** [comment]
I'd rather we avoid further polluting JS expression space with new expression-level operators. Could we achieve this with a magic type similar to `ThisType<T>`? 

```js
// NOTE: Follows new mechanism for defining magic types like `Uppercase<T>` 
// as opposed to `ThisType<T>`
type Satisfies<T> = intrinsic; 
 
const options = { } as Satisfies<Options>;
```

**Comment by andrewbranch** [comment]
I find the postfix operation nice for the “ensure this literal I wrote in the middle of some expression is assignable to `T`,” but very awkward for flowing contextual types into the expression, which is a case I find more compelling on the whole. Consider

```ts
const actions = {
  foo: payload => {
    // ...
  },
  bar: payload => {
    // ...
  },
  baz: payload => {
    // ...
  },
  one: payload => {
    // ...
  },
  fish: payload => {
    // ...
  },
} satisfies Record<string, (payload: Whatever) => Promise<boolean>>;
```

You had to do a lot of reading before finding out what each `payload` parameter was supposed to be, and I omitted the body of each function. Even when you get there, the flowing lexically backwards of the contextual type feels unexpected to me. Of the syntax proposals I’ve seen, I kind of like `as extends T` because it feels like it could naturally be extended to variable declarations:

```ts
const actions: extends Record<string, (payload: Whatever) => Promise<boolean>> = {
  // ...
}
```

That does have the issue that an `extends T` annotation is meaningless in absence of an initializer and would probably have to be a grammar error, so it’s far from perfect, but having some way of writing the contextual type _before_ its expression would feel much more ergonomic to me.

**Comment by RyanCavanaugh** [comment]
I gave a lot of examples involving `const x: T = ...`-like constructs but I think a lot of the use cases here are in expression positions where there's no place to put a type annotation (e.g. function argument). Anything we do here should be useable in any expression position IMO

**Comment by awerlogus** [comment]
What about jsdoc compatibility?

**Comment by ChiriVulpes** [comment]
> I kind of like as extends T because it feels like it could naturally be extended to variable declarations: [...] having some way of writing the contextual type before its expression would feel much more ergonomic to me.

> I gave a lot of examples involving `const x: T = ...`-like constructs but I think a lot of the use cases here are in expression positions where there's no place to put a type annotation (e.g. function argument). Anything we do here should be useable in any expression position IMO

Considering that this syntax is filling a similar role as a type assertion, it makes sense to me that whatever keyword is used can be positioned both as a postfix and part of declaring the type of a variable, to mirror type assertions. And, like... couldn't both positions be supported with either of these keywords? Like, all the the below examples seem equally sensible to me in reading them. Am I misunderstanding why not all of these examples could work?
```ts
const foo: satisfies Bar = { ... };
applyBar({ ... } satisfies Bar);
```

```ts
const foo: extends Bar = { ... };
applyBar({ ... } as extends Bar);
```

Note that I personally cast no vote about which syntax to use, as I like the sound of "satisfies", and "extends" is kinda all over the place in TypeScript already, filling a ton of different roles, which can make code a little bit slower to comprehend imo. But at the same time, being able to use `as` in the postfix for this mirrors type assertions better, so... No preference here. The only camp I'm in is the camp of having both a type declaration and a postfix version of this operator. 😜

**Comment by danvk** [comment]
Whatever the syntax winds up being, it would be great if it could also be used to give a contextual type to JSON imports (#37831):

```ts
import data from './data.json' satisfies MyType;
```

**Comment by treybrisbane** [comment]
Got a question on this.

Given
```ts
interface Foo {
  bar: 1 | 2;
}

const value = { bar: 1 as const } satisfies Foo;
```
What is the type of `value`?

**Comment by ChiriVulpes** [comment]
@treybrisbane 
> What is the type of `value`?

```ts
{ bar: 1 }
```
It's not a postfix to change the type, it's a postfix to *validate* the type. It's making sure whatever value is there has a type which, in this case, is assignable to `Foo`

**Comment by treybrisbane** [comment]
>I'm pretty sure there's some other stuff going on with like, turning parts of the value's type into the type in satisfies via contextual typing, most likely exclusively in cases that don't widen it. I might be wrong on that front though? That's stuff I don't really understand besides how to use it haha

That's pretty much exactly the stuff I'm asking about. 😄
Although in hindsight, my example was maybe a bit too simple to dig into it properly. 🤔

**Comment by jmroon** [comment]
This will be incredibly useful for using online types with the data object literal in Vue. Looking forward to this!

**Comment by webstrand** [comment]
Can we have
```ts
foo.bar satisfies string;
```
be completely erased?
Right now it leaves behind the property access in the compiled javascript code.

**Comment by fatcerberus** [comment]
@webstrand TS can't safely erase that because `foo.bar` might trigger a getter call and the type system doesn't generally distinguish between accessors and fields.

**Comment by RyanCavanaugh** [comment]
[Discussion](https://github.com/microsoft/TypeScript/issues/7481#issuecomment-1012386203)

**Comment by RyanCavanaugh** [comment]
This example is interesting:
```ts
let obj: { f(s: string): void } & Record<string, unknown> = {
    f(s) { },
    g(s) { }
} satisfies { g(s: string): void } & Record<string, unknown>;
```
When an `e satisfies T` expression is contextually typed by a type `U`, it looks like we need to contextually type `e` by `T & U`

**Comment by RyanCavanaugh** [comment]
Ref. #47920

**Comment by RyanCavanaugh** [comment]
@typescript-bot pack this

**Comment by RyanCavanaugh** [comment]
Merged up to main, added outer&inner intersectional contextual type behavior, and added another test based on the issue discussion ✔

File: src/compiler/checker.ts
Comment (RyanCavanaugh) [reply src/compiler/checker.ts:27018]: My brain lost count of the parent cardinality and was trying to guard against calling `getContextualType` on a statement. Fixing.

File: tests/baselines/reference/satisfiesScenarios.errors.txt
Comment (DanielRosenwasser) [inline tests/baselines/reference/satisfiesScenarios.errors.txt:16]: Nit - I would place this test under `conformance`.

File: tests/cases/compiler/satisfiesScenarios.ts
Comment (DanielRosenwasser) [inline tests/cases/compiler/satisfiesScenarios.ts:112]: Can we add some tests for cases where the intersection of the expected type and the contextual type is vacuous?

For example

```ts
namespace VacuousIntersectionOfContextualTypes {
    const a: "baz" = "foo" satisfies "foo" | "bar";

    const b: { xyz: "baz" } = { xyz: "foo" } satisfies { xyz: "foo" | "bar" };
}

```

File: tests/cases/compiler/satisfiesScenarios.ts
Comment (DanielRosenwasser) [inline tests/cases/compiler/satisfiesScenarios.ts:55]:
```suggestion
    // Should fail under --noPropertyAccessFromIndexSignature
```

File: tests/cases/compiler/satisfiesScenarios.ts
Comment (DanielRosenwasser) [inline tests/cases/compiler/satisfiesScenarios.ts:58]: Do you have a toggled version of this test at all to test these scenarios?

File: tests/cases/compiler/satisfiesScenarios.ts
Comment (DanielRosenwasser) [reply tests/cases/compiler/satisfiesScenarios.ts:112]: Also, maybe a test that combines both inference and contextual typing?

**Review by DanielRosenwasser — CHANGES_REQUESTED** [review]
I think we need more test coverage.

File: tests/baselines/reference/satisfiesScenarios.errors.txt
Comment (a-tarasyuk) [reply tests/baselines/reference/satisfiesScenarios.errors.txt]: I've moved to conformance

File: tests/cases/compiler/satisfiesScenarios.ts
Comment (a-tarasyuk) [reply tests/cases/compiler/satisfiesScenarios.ts]: I think call doesn't directly affect `satisfies` usage. Anyway I've added two tests `typeSatisfaction_propertyValueConformance1`, `typeSatisfaction_propertyValueConformance1` with `noPropertyAccessFromIndexSignature: true, false` and `@noUncheckedIndexedAccess: true, false`

**Comment by Jack-Works** [comment]
Tried this, have some feedback:

image

`a satisfies T` looks like `: T = a`, so the error should happen on the value, not the type constraint.

image

`satisfies` should also be available on FunctionDeclaration.

**Comment by mon-jai** [comment]
> Tried this, have some feedback:
> 
> image
> 
> `a satisfies T` looks like `: T = a`, so the error should happen on the value, not the type constraint.
> 
> image
> 
> `satisfies` should also be available on FunctionDeclaration.

It would be really useful for React developers.

```jsx
import React, { FC } from "react";

type GreetingProps = {
  name: string;
}

function Greeting ({ name }) => {
  // name is string!
  return <h1>Hello {name}</h1>
} satisfies FC<GreetingProps>
```

**Comment by kasperpeulen** [comment]
What is the status of this PR? @a-tarasyuk

**Comment by a-tarasyuk** [comment]
@kasperpeulen It's up to date 😄. Looks like #47920 is still in discussion. I don't know about the current status because I'm not part of the TypeScript team. So I think it would be better to ask @RyanCavanaugh, maybe he can provide some updates.

**Comment by RyanCavanaugh** [comment]
We're ready to merge this for 4.9

File: src/compiler/checker.ts
Comment (RyanCavanaugh) [inline src/compiler/checker.ts:27406]: After discussion in the original issue I think we want to just use the asserted type, so this can just be
```suggestion
                    return getTypeFromTypeNode((parent as SatisfiesExpression).type);
```

File: src/compiler/checker.ts
Comment (RyanCavanaugh) [inline src/compiler/checker.ts:32333]:
```suggestion
```

**Review by RyanCavanaugh — APPROVED** [review]
LGTM after a couple edits + merge conflicts

**Comment by mon-jai** [comment]
> satisfies should also be available on FunctionDeclaration.

Will you implement it? @RyanCavanaugh

File: src/compiler/checker.ts
Comment (a-tarasyuk) [reply src/compiler/checker.ts]: Ok

**Review by DanielRosenwasser — APPROVED** [review]

**Comment by Jack-Works** [comment]
> > `satisfies` should also be available on FunctionDeclaration.
> 
> It would be really useful for React developers.
> 
> ```js
> import React, { FC } from "react";
> 
> function Greeting<T>(props: { name: string, data: T }) => {
>   return ...
> } satisfies React.ComponentType<any>
> ```

cc @RyanCavanaugh what about this?

**Comment by WORMSS** [comment]
Just confirming that type error codes on errors are expected to swap?

**Comment by leerob** [comment]
Extremely excited about this for Next.js: https://twitter.com/leeerob/status/1563540593003106306

```ts
import { GetStaticProps, InferGetStaticPropsType } from "next";

export const getStaticProps = () => {
  return {
    props: {
      id: "123",
    },
  };
} satisfies GetStaticProps;

export default function Page(
  props: InferGetStaticPropsType<typeof getStaticProps>,
) {
  props;
  // ^?(parameter) props: { id: string;}
}
```

**Comment by texastoland** [comment]
Was there any discussion about repurposing `implements` instead of introducing `satisfies` to reduce conceptual load? Does `satisfies` have precedent in a language I'm not familiar with? I think #47920 raises worthwhile points at least to explain or link how it came to be. Love what you all do 👏🏼

**Comment by andrewbranch** [comment]
@texastoland yes, lots of syntax alternatives, including `implements`, were discussed. Some notes about `implements` were captured: https://github.com/microsoft/TypeScript/issues?q=is%3Aissue+label%3A%22Design+Notes%22+satisfies+implements+is%3Aclosed+

**Comment by cefn** [comment]
Is there a possibility of behaviour like https://tsplay.dev/mLyvAW being reversed now that there is a way to determine whether you want a reference to have the original type (just use satisfies) or the broadened type (assign to a typed variable)? 

Or would that break too much?

To see the behaviour that I found surprising hover over `basicValue` vs `castValue`. I expected the explicit nature of the variable type declaration to be respected, but it actually needs a cast to force the issue which is pretty hard to explain to people. I can see how this was perhaps unavoidable before `satisfies` but maybe now...?

**Comment by noinkling** [comment]
> What about jsdoc compatibility?

Quoting this since it seems it was never addressed. Any chance of a JSdoc equivalent?

The equivalent for `as` is `/** @type {SomeType} */(someExpression)`, for reference.

**Comment by andrewbranch** [comment]
@noinkling mind opening a new issue for tracking?

**Comment by a-tarasyuk** [comment]
@noinkling I think a new issue would be helpful, at least to move the discussion there. @andrewbranch I'm not sure what tag we can use for this https://jsdoc.app/index.html

**Comment by andrewbranch** [comment]
It was discussed in a design meeting (#50522), but I don’t think we had a conclusion:

> - JSDoc?
>   - We try not to add new tags that aren't doc'd in the usejsdoc site.
>   - Ship has sailed with TS-specific syntax in JSDoc, right?

**Comment by brendankenny** [comment]
Opened #51086 for satisfied in jsdoc discussion

**Comment by Andarist** [comment]
I wonder if it was a deliberate decision to preserve the "original" type of expression for expressions that are not satisfied by the operator (in case of an error). This is different behavior from the `as` operator:
[TS playground](https://www.typescriptlang.org/play?ts=4.9.0-dev.20221011#code/JYWwDg9gTgLgBAbwOIFMYGUVQG5fcAExQAUoIwBnAXzgDMyQ4ByAOxQA8YmAobgYwgsK8AOZpMOPIRJlKcALxwAFAEMKATxZ9lASgUA+ZQjhhZFAFyI4MFMMsBGODSo69FFTGAVawW3FQYWLhQ+ESk5BQA3NwA9DFwCXAAegD8vAJCouJBUmFmAEwKykp68oYILnDunt6+FP7ZkiHS4ZTRcYnJafyCwnBigU2hMhEAzEVKJQaIlWoNg8HDrVGx8YmpQA)

**Comment by RyanCavanaugh** [comment]
We didn't discuss what to do in error cases, oddly enough. I think keeping the expression type regardless of outcome is the most consistent/predictable behavior, even if the *net* behavior is maybe a little less desirable (since, presumably, you now have two errors in your program instead of one).

I think there's something weird going on with the language service displaying the asserted type in the Playground. The actual type of the variable, shown in .d.ts emit, is the expression type.

**Comment by edisonLzy** [comment]
> 

maybe { bar: 1 } ?

**Pull request merged**
